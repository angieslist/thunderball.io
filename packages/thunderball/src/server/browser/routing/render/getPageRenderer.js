import React from 'react';
import Helmet from 'react-helmet';
import NodeCache from 'node-cache';
import Shell from 'thunderball-client/lib/render/Shell';
import serialize from 'serialize-javascript';
import _ from 'lodash';
import { renderToString, renderToNodeStream } from 'react-dom/server';
import { Stream } from 'stream';
import memoize from 'memoizee';
import constants from '../../../../constants';

const MEMOIZE_MAX_AGE = _.get(constants.APP_CONFIG, 'ssr.caching.memoizeMaxAge', 3600000);

const intlPolyfillFeatures = _.get(constants.APP_CONFIG, 'i18n.locales', [])
  .map(locale => `Intl.~locale.${locale}`)
  .join();

// Allow custom scripts from application config and ion manifest
// If any scripts are a function, execute them, otherwise just display them
const getCustomScripts = (page: Object, scriptType: String): Array<String> => [
  ..._.get(constants.APP_CONFIG, `ssr.${scriptType}`, []),
  ..._.get(page, `ssr.${scriptType}`, []),
].map(script => ((typeof script === 'function') ? script() : script))
  .filter(script => !!script); // remove any falsey scripts

export const getHtmlProperties = (page, name) => {
  const helmet = Helmet.rewind();
  const { styles, javascript } = webpackIsomorphicTools.assets();

  return {
    htmlAttrs: helmet.htmlAttributes.toString(),
    bodyAttrs: helmet.bodyAttributes ? helmet.bodyAttributes.toString() : '',
    head: [
      helmet.title.toString(),
      helmet.base.toString(),
      helmet.meta.toString(),
      helmet.link.toString(),
      helmet.style.toString(),
      helmet.script.toString(),
      helmet.noscript.toString(),
      // Get the common css file and the css file for the page
      ...[styles.commons, styles[name]].filter(v => !!v).map((v, k) => `<link href='${v}' rel='stylesheet' key='${k}' />`),
    ].join(''),
    scripts: {
      beforeBody: getCustomScripts(page, 'beforeBody'),
      afterBody: [
        ...[
          `<script type="text/javascript" src="https://cdn.polyfill.io/v2/polyfill.min.js?features=${intlPolyfillFeatures}"></script>`,
          javascript.commons ? `<script type="text/javascript" src="${javascript.commons}"></script>` : '',
          `<script type="text/javascript" src="${javascript[name]}"></script>`,
        ],
        ...getCustomScripts(page, 'afterBody'),
      ],
    },
  };
};
const memoizedGetHtmlProperties = memoize(getHtmlProperties,
  { maxAge: MEMOIZE_MAX_AGE, preFetch: true, normalizer: args => args[2] });

const htmlCache = new NodeCache({ stdTTL: MEMOIZE_MAX_AGE / 1000, errorOnMissing: true });

// Node stream that caches data that flows through it.
const createCacheStream = (cacheKey) => {
  const bufferedChunks = [];
  return new Stream.Transform({
    transform(data, enc, cb) {
      bufferedChunks.push(data);
      cb(null, data);
    },
    flush(cb) {
      htmlCache.set(cacheKey, Buffer.concat(bufferedChunks));
      cb();
    },
  });
};

const createCachedRenderer = (preShell, appShell, postShell, cacheKey) => ({
  toPromise() {
    try {
      const cachedHtml = htmlCache.get(cacheKey);
      return Promise.resolve(cachedHtml);
    } catch (err) {
      return new Promise((resolve) => {
        const renderedPage = `${preShell}${renderToString(appShell)}${postShell}`;
        htmlCache.set(cacheKey, renderedPage);
        resolve(renderedPage);
      });
    }
  },
  toStream() {
    try {
      const cachedHtml = htmlCache.get(cacheKey);
      const passThrough = new Stream.PassThrough();
      passThrough.end(cachedHtml);
      return passThrough;
    } catch (err) {
      const cacheStream = createCacheStream(cacheKey);
      const reactStream = renderToNodeStream(appShell);
      cacheStream.write(preShell, () => reactStream.pipe(cacheStream, { end: false }));
      reactStream.on('end', () => cacheStream.end(postShell));
      return cacheStream;
    }
  },
});

const getPageRenderer = ({
  store,
  renderProps,
  req,
  page,
  name,
  cacheKey,
  reactRendererCacheKey,
  ssrConfig,
  flushCache = false,
}) => {
  if (flushCache) {
    htmlCache.flushAll();
  }

  const state = store.getState();

  // Render the application body only if page.ssr.renderBody is not 'false'
  const appBody = _.get(ssrConfig, 'renderBody', true) !== false ?
    <Shell {...{ store, renderProps }} pageProps={page.pageProps} defaultLocale={_.get(constants.APP_CONFIG, 'i18n.defaultLocale')} />
    : '';

  const shouldMemoize = constants.IS_PRODUCTION && (_.get(ssrConfig, 'caching.memoize', true) !== false);

  // Determine html properties, they can be memoized except for the `state` which could change
  const html = (cacheKey && shouldMemoize ? memoizedGetHtmlProperties : getHtmlProperties)(page, name, cacheKey);

  if (!constants.IS_PRODUCTION) {
    webpackIsomorphicTools.refresh();
  }

  const preShell = `<html ${html.htmlAttrs}><head>${html.head}</head><body ${html.bodyAttrs}><script type="text/javascript">window.__INITIAL_STATE__ = ${serialize(state)};</script>${html.scripts.beforeBody.join('')}<div id="app">`;
  const postShell = `</div>${html.scripts.afterBody.join('')}</body></html>`;

  const cacheKeyStr = `${name}:${JSON.stringify(reactRendererCacheKey)}`;

  return createCachedRenderer(preShell, appBody, postShell, cacheKeyStr);
};

export default getPageRenderer;
