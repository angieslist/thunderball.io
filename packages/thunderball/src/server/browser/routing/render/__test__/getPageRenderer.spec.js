import Helmet from 'react-helmet-async';
import { Stream } from 'stream';
import getPageRenderer, { getHtmlProperties } from '../getPageRenderer';

const streamToPromise = stream => (new Promise((resolve, reject) => {
  let buffer = '';
  stream.on('data', (data) => { buffer += data; });
  stream.on('end', () => resolve(buffer));
  stream.on('error', () => reject());
}));

describe('getPageRenderer.js', () => {
  let webpackIsomorphicTools;
  beforeAll(() => {
    // Let Helmet know we want to test this as if on the server
    // (see https://github.com/nfl/react-helmet/blob/master/test/HelmetTest.js)
    Helmet.canUseDOM = false;
    // stub global webpackIsomorphicTools
    // TODO make functions in app code pure and eventually get rid of globals
    webpackIsomorphicTools = global.webpackIsomorphicTools;
    global.webpackIsomorphicTools = {
      assets: () => ({
        styles: {
          commons: 'commonStyle',
          testPage: 'testStyle',
        },
        javascript: {
          commons: 'commonJs',
          testPage: 'testJs',
        },
      }),
      refresh: () => {},
    };
  });
  afterAll(() => {
    Helmet.canUseDOM = true;
    // replace stub with original global value
    global.webpackIsomorphicTools = webpackIsomorphicTools;
  });
  describe('getPageRenderer', () => {
    const store = {
      getState: () => {},
    };
    const renderProps = {};
    const req = {};
    const page = {};
    const name = 'testPage';
    const cacheKey = null;
    const reactRendererCacheKey = null;
    const ssrConfig = null;
    it('returns a renderer', async () => {
      const pageRenderer = getPageRenderer({
        store,
        renderProps,
        req,
        page,
        name,
        cacheKey,
        reactRendererCacheKey,
        ssrConfig,
        helmetContext: {
          helmet: {
            htmlAttributes: '',
            title: '',
            base: '',
            meta: '',
            link: '',
            style: '',
            script: '',
            noscript: '',
          },
        },
      });

      expect(pageRenderer.toPromise).toBeInstanceOf(Function);
      expect(pageRenderer.toStream).toBeInstanceOf(Function);
    });
    it('renders html to a promise', async () => {
      const pageRenderer = getPageRenderer({
        store,
        renderProps,
        req,
        page,
        name,
        cacheKey,
        reactRendererCacheKey,
        ssrConfig,
        flushCache: true,
        helmetContext: {
          helmet: {
            htmlAttributes: '',
            bodyAttributes: '',
            title: '',
            base: '',
            meta: '',
            link: '',
            style: '',
            script: '',
            noscript: '',
          },
        },
      });

      const html = await pageRenderer.toPromise();
      expect(html).toMatchSnapshot();
      const cachedHtml = await pageRenderer.toPromise();
      expect(cachedHtml).toMatchSnapshot();
    });
    it('renders html to a node stream', async () => {
      const pageRenderer = getPageRenderer({
        store,
        renderProps,
        req,
        page,
        name,
        cacheKey,
        reactRendererCacheKey,
        ssrConfig,
        flushCache: true,
        helmetContext: {
          helmet: {
            htmlAttributes: '',
            bodyAttributes: '',
            title: '',
            base: '',
            meta: '',
            link: '',
            style: '',
            script: '',
            noscript: '',
          },
        },
      });

      const stream = pageRenderer.toStream();
      expect(stream).toBeInstanceOf(Stream.Readable);
      const html = await streamToPromise(stream);
      expect(html).toMatchSnapshot();
      const cachedStream = pageRenderer.toStream();
      const cachedHtml = await streamToPromise(cachedStream);
      expect(cachedHtml).toMatchSnapshot();
    });
  });
  describe('getHtmlProperties', () => {
    it('gets HTML properties', () => {
      const htmlProperties = getHtmlProperties(
        {},
        'testPage',
        null,
        {
          helmet: {
            htmlAttributes: '',
            bodyAttributes: '',
            title: '',
            base: '',
            meta: '',
            link: '',
            style: '',
            script: '',
            noscript: '',
          },
        },
      );
      expect(htmlProperties).toMatchSnapshot();
    });
  });
});
