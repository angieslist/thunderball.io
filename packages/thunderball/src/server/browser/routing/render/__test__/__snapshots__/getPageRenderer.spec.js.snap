// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getPageRenderer.js getHtmlProperties gets HTML properties 1`] = `
Object {
  "bodyAttrs": "",
  "head": "<title data-react-helmet=\\"true\\"></title><link href='commonStyle' rel='stylesheet' key='0' /><link href='testStyle' rel='stylesheet' key='1' />",
  "htmlAttrs": "",
  "scripts": Object {
    "afterBody": Array [
      "<script type=\\"text/javascript\\" src=\\"https://cdn.polyfill.io/v2/polyfill.min.js?features=\\"></script>",
      "<script type=\\"text/javascript\\" src=\\"commonJs\\"></script>",
      "<script type=\\"text/javascript\\" src=\\"testJs\\"></script>",
    ],
    "beforeBody": Array [],
  },
}
`;

exports[`getPageRenderer.js getPageRenderer returns rapscallion Renderer 1`] = `"<!DOCTYPE html data-react-checksum=\\"-944392483\\"><html ><head><title data-react-helmet=\\"true\\"></title><link href='commonStyle' rel='stylesheet' key='0' /><link href='testStyle' rel='stylesheet' key='1' /></head><body ><script type=\\"text/javascript\\">window.__INITIAL_STATE__ = undefined;</script><div id=\\"app\\"><!-- react-empty: 1 --></div><script type=\\"text/javascript\\" src=\\"https://cdn.polyfill.io/v2/polyfill.min.js?features=\\"></script><script type=\\"text/javascript\\" src=\\"commonJs\\"></script><script type=\\"text/javascript\\" src=\\"testJs\\"></script></body></html>"`;
